rfd/0149-headless-kube.md

```diff
@@ -0,0 +1,257 @@
+are separated from each other, in case the remote machine is shared by multiple people.
+
+# Details
+
+To implement headless Kubernetes access, we'll build upon two recently implemented features -
+headless authentication for `tsh ssh` ([RFD 105](https://github.com/gravitational/teleport/blob/master/rfd/0105-headless-authentication.md)) and Teleport Kubernetes
+local proxy ([RFD 121](https://github.com/gravitational/teleport/blob/master/rfd/0121-kube-mfa-sessions.md)).
+
+## Headless authentication
+
+When headless authentication is initiated by the user on a remote machine, the client calls the `POST /webapi/login/headless`
+endpoint on the Auth server. It creates a `HeadlessAuthentication` resource on the backend.
+After that, the call is paused, and the user is provided with a link (or command line alternative) where they can approve
+the newly created headless authentication request using their local machine with a logged-in Teleport and an MFA device. Once
+the request is approved by the user, the authentication process is unpaused and the original call returns regular
+user's Teleport login certificates created by the auth server. These certificates have a TTL of 1 minute.
+
+## Headless Kubernetes proxy
+
+We will add the flag `--headless` to the `tsh proxy kube` command.
+The user will be able to start `tsh --headless proxy kube` in headless mode on a remote machine,
+then authenticate on the local machine using the link provided by the started proxy (using the flow
+described in the previous section).
+
+After the user finishes the authentication process, the headless kube proxy will generate certificates for accessing
+specified Kubernetes clusters. When the Auth server issues these certificates, it will recognize that
+the certificate is being generated for the user kube proxy - certificate usage is `UsageKubeOnly` and
+`RequesterName` is `TSH_KUBE_LOCAL_PROXY`. For this case, it will not limit the certificate TTL by the calling
+identity TTL, which is 1 minute, since that's what was created during headless login. Instead, these certificates'
+expiration will be limited by the `max_session_ttl` time, similar to how the regular `tsh proxy kube` works.
+When the kube proxy receives certificates for accessing Kubernetes clusters, it will create an in-memory file and
+will write kubeconfig to it. This kubeconfig is only needed for correct access from the kubectl to the local kube proxy;
+it does not contain certificates information for accessing the Kubernetes cluster in teleport. To create the kubeconfig we
+generate a new one-use local key every time the proxy starts. Next, the headless kube proxy will reexec itself, opening a new
+shell for the user, while keep running in the background. The kube proxy will provide this new shell with the
+access to the in-memory kubeconfig by the path `/proc/self/fd/3` and the proxy will point the `$KUBECONFIG`
```

Comment:
Just curious, what would be the behavior if you use terminal multiplexer (tmux, screen) which is pretty common on shared bastions. It's a separate process so it won't be shared between your host bash and, say, screen session?

Reply:
I don't think there should be problems. I've tested it with tmux (SSHed into a machine and ran tmux there) and it seems to be working completely fine. Here's how processes tree looks like for tmux:
![image](https://github.com/gravitational/teleport/assets/964862/d9c27884-d11c-4bca-9a69-7342eef73eb5)